useless update_appearance reduction, emissive_blocker micro optimization (saves a second of init)

LemonInTheDark opened the pull request:
## About The Pull Request

[Saves 0.2 seconds of init time. 50% of emissive blockers](https://github.com/tgstation/tgstation/commit/8318b648f6d32844aacbfb4c309152cd45801f5c)

Emissive blockers are a decent expense during init, even these, which are the ones that update outside of initialize.
I've inlined them, removed some redundant vars and checks, reduced the arg count, and shifted some things around. This ends up saving 200ms, or 50% of its total cost.

I also shifted mutable_appearance about a bit. it's not a massive saving, but it is technically faster

[Prevents a few redundant appearance_updates, saves 0.8 seconds of init](https://github.com/tgstation/tgstation/commit/5475cd778b66b22b1e2c8d86b2c6d59fb84f219a)

Prequisit info: update_appearance is decently expensive
It's good then to only do it if we have a reason to, right?

Me and moth were shooting the shit about just general init time, and we came up with the idea of tracking which update_appearances actually "worked" and which didn't.

That bit comes later, let's enjoy the fruits of that work first

First, holograms were calling update_appearance on process, for almost no reason.
I patched the one event they don't already "react" to, and then locked it behind a change decection if.
good for live, doesn't impact init.

Next, decals. If you add a decal to something before it inits, it'll react to the after successful init signal.
The trouble is the same atom could have its appearance updated from this MORE then once, since decals can be stacked on tiles, and signal unregisters don't work once the signal is sent.
So we add a flag to track if we've had this happen to us or not, so it only happens once.
saves 80 ms

Power! lots of things call power_change on init, often more then once.
We'll update appearance for each of those calls, even if only one is an actual change.
That's silly, better to track what sort of power we're using for our appearance and go off that changing

This was taking about 300ms. Really stupid

Icon smoothing. After emissive blockers were added, any change to something's icon smoothing would lead to an update_appearance call.
Nasty shit, specially cause of walls, which don't even use emissive blockers.
Ok then, so we'll always update appearance for movables, and will allow turfs that are interested to hook it manually.
Not many of those anyhow
This is slightly a dev ux thing, but it saves 600ms so I think it's worth it. Rare case anyway

Telecomms:
telecomm machines were updating appearance on process. This is to cover for them turning on/off on process.
Better then to just check if on actually changed.
This cost adds up midgame, doesn't impact init tho

Materials:
There's this update_appearance call in material on_apply. it doesn't do anything.
The logs will lie to you and say it does, but it's just like reapplying emissives. It doesn't need to exist
Saves like 50ms

Canisters:
Live thing, lots of time wasted updating appearance for no reason, lets see if we change anything first yes?

[Uses defines to wrap update_appearance for tracking](https://github.com/tgstation/tgstation/commit/4fa82e1c9d93577aadb3c743f17196331f62e67c)

[Undoes _update_appearance changes, instead reccomends 2 regexes to use](https://github.com/tgstation/tgstation/commit/a8c8fec57a4e43d1fa636b5ac68459903faa9fc5)

I need file and line number for my tracking, so I need to override update_appearance calls, and also preferably not require every override of update_appearance to handle dummy file + line args.

So instead, I created a wrapper proc that checks to see if appearanaces match (they're unique remember, the two of the same visual appearance will be equivalent)
The trouble is I can't intercept JUST proc calls, or JUST function definitions with defines. it needs to be both.

So I renamed the /update_appearance proc to /_update_appearance

this way I can capture old uses, and don't need to worry about merge/dev brain skew

~~It does mean that all update_appearance proc definitions now look weird tho.
My profiling is leaking into dev ux. I wish I had better templating.~~

**The above is no longer being pr'd**, it's instead just recommended via a few regexes adjacent to the define. 
Smelled wrong anyhow

[Adds a setter for panel_open, so I can update_appearance on it](https://github.com/tgstation/tgstation/pull/71658/commits/cf1df8a69fc1a816391d085ee7419b14f9fe9167)

## Why It's Good For The Game

Speed

LemonInTheDark committed:
Saves 0.2 seconds of init time. 50% of emissive blockers

Emissive blockers are a decent expense during init, even these, which
are the ones that update outside of initialize.
I've inlined them, removed some redundant vars and checks, reduced the arg count,
and shifted some things around. This ends up saving 200ms, or 50% of its
total cost.

I also shifted mutable_appearance about a bit. it's not a massive
saving, but it is technically faster

LemonInTheDark committed:
Prevents a few redundant appearance_updates, saves 0.8 seconds of init

Prequisit info: update_appearance is decently expensive
It's good then to only do it if we have a reason to, right?

Me and moth were shooting the shit about just general init time, and we
came up with the idea of tracking which update_appearances actually
"worked" and which didn't.

That bit comes later, let's enjoy the fruits of that work first

First, holograms were calling update_appearance on process, for almost
no reason. I patched the one event they don't already "react" to, and
then locked it behind a change decection if. good for live, doesn't
impact init.

Next, decals. If you add a decal to something before it inits, it'll
react to the after successful init signal. The trouble is the same atom
could have its appearance updated from this MORE then once, since decals
can be stacked on tiles, and signal unregisters don't work once the
signal is sent.
So we add a flag to track if we've had this happen to us or not, so it
only happens once. saves 80 ms

Power! lots of things call power_change on init, often more then once.
We'll update appearance for each of those calls, even if only one is an actual change.
That's silly, better to track what sort of power we're using for our
appearance and go off that changing

This was taking about 300ms. Really stupid

Icon smoothing. After emissive blockers were added, any change to
something's icon smoothing would lead to an update_appearance call.
Nasty shit, specially cause of walls, which don't even use emissive
blockers.

Ok then, so we'll always update appearance for movables, and will
allow turfs that are interested to hook it manually. Not many of those
anyhow
This is slightly a dev ux thing, but it saves 600ms so I think it's
worth it. Rare case anyway

Telecomms:
telecomm machines were updating appearance on process. This is to cover
for them turning on/off on process.
Better then to just check if on actually changed. This cost adds up
midgame, doesn't impact init tho

Materials:
There's this update_appearance call in material on_apply. it doesn't do
anything.
The logs will lie to you and say it does, but it's just like reapplying
emissives. It doesn't need to exist

LemonInTheDark committed:
Uses defines to wrap update_appearance for tracking

I need file and line number for my tracking, so I need to override
update_appearance calls, and also preferably not require every override
of update_appearance to handle dummy file + line args.

So instead, I created a wrapper proc that checks to see if appearanaces
match (they're unique remember, the two of the same visual appearance will be equivilant)
The trouble is I can't intercpet JUST proc calls, or JUST function
definitions with defines. it needs to be both.

So I renamed the /update_appearance proc to /_update_appearance

this way I can capture old uses, and don't need to worry about merge/dev
brain squew

It does mean that all update_appearance proc definitions now look weird
tho. My profiling is leaking into dev ux. I wish I had better
templating.

LemonInTheDark committed:
Undoes _update_appearance changes, instead reccomends 2 regexes to use

LemonInTheDark commented:
#71645 - the _update_appearance bit. will tm that, this is fit for review (assuming screenshots don't fail again)

LemonInTheDark committed:
sanitizes csv

LemonInTheDark committed:
makes canisters hold off on visual updates unless something isactually changing

LemonInTheDark committed:
fuck

Mothblocks commented:
If those don't show up in game like that, then it might be exposing a bug in getFlatIcon itself with these changes? There's obviously *a* bug.

Mothblocks commented:
Didn't mean to hide the other one I thought there was more

LemonInTheDark committed:
Fixed the funny haha black overlays

I forgot to have the "no turf exists" case default to the base plane. am
stupid, sorry :(

LemonInTheDark commented:
Comeon gimme the good roll daddy needs some cpu time

LemonInTheDark committed:
Fixes up a miscalled update_appearance, fixes mapping helper'd airlocks not looking right on mapload

Kylerace commented on code/__HELPERS/lighting.dm:
```diff
@@ -1,9 +1,29 @@
 /// Produces a mutable appearance glued to the [EMISSIVE_PLANE] dyed to be the [EMISSIVE_COLOR].
 /proc/emissive_appearance(icon, icon_state = "", atom/offset_spokesman, layer = FLOAT_LAYER, alpha = 255, appearance_flags = NONE, offset_const)
```
you still have the alpha argument

LemonInTheDark replied:
part of me thinks it could still be implemented, and I don't have the guts to convince myself otherwise

Kylerace commented on code/__HELPERS/lighting.dm:
```diff
@@ -1,9 +1,29 @@
+/proc/fast_emissive_blocker(atom/make_blocker)
+	// Note: alpha doesn't "do" anything, since it's overriden by the color set shortly after
+	// Consider removing it someday?
+	var/mutable_appearance/MA = new()
```
it should clone the atoms appearance and then set relevant blocker specific vars

Mothblocks replied:
better name

Kylerace commented on code/__HELPERS/stat_tracking.dm:
```diff
@@ -1,3 +1,4 @@
+// For use with the stopwatch defines
```
```suggestion
/// For use with the stopwatch defines
```

Kylerace commented on code/__HELPERS/stat_tracking.dm:
```diff
@@ -12,7 +13,8 @@
 	. = lines.Join("\n")
 
-/proc/stat_tracking_export_to_file_later(filename, costs, counts)
+// For use with the set_cost defines
```
```suggestion
/// For use with the set_cost defines
```

LemonInTheDark replied:
similar, //////////// EXCITED GROUPS //////////// is the language of my fathers

Kylerace commented on code/datums/elements/decals/_decal.dm:
```diff
@@ -128,11 +128,11 @@
 	SIGNAL_HANDLER
 
-	if(source && istype(source))
+	if(istype(source) && !(source.flags_1 & DECAL_INIT_UPDATE_EXPERIENCED_1))
```
babe wake up new flags_1 just dropped

LemonInTheDark replied:
ye

Kylerace commented on code/datums/materials/_material.dm:
```diff
@@ -92,8 +92,6 @@ Simple datum which is instanced once per type and is used for every object of sa
 	else if(istype(source, /turf)) //turfs
 		on_applied_turf(source, amount, material_flags)
 
-	source.update_appearance()
```
?

Kylerace commented on code/game/machinery/telecomms/telecomunications.dm:
```diff
@@ -153,13 +153,12 @@ GLOBAL_LIST_EMPTY(telecomms_list)
 /obj/machinery/telecomms/process(delta_time)
 	update_power()
 
-	// Update the icon
```
the appearance also relies on panel_open

LemonInTheDark replied:
good point, implemented a setter

Mothblocks commented on code/datums/mutable_appearance.dm:
```diff
@@ -24,27 +24,25 @@
  * offset_const - A constant to offset our plane by, so it renders on the right "z layer"
 **/
 /proc/mutable_appearance(icon, icon_state = "", layer = FLOAT_LAYER, atom/offset_spokesman, plane = FLOAT_PLANE, alpha = 255, appearance_flags = NONE, offset_const)
+	var/mutable_appearance/MA = new()
```
better name

Mothblocks commented on code/modules/power/power.dm:
```diff
@@ -426,6 +433,6 @@
 	for(var/obj/structure/cable/C in src)
 		if(C.machinery_layer & machinery_layer)
-			C.update_appearance()
+			C.update_appearance() // I hate this. it's here because update_icon_state SCANS nearby turfs for objects to connect to. Wastes cpu time
```
fwiw i think you should remove this comment because my PR is coming soon and it mostly fixes the problematic part of this

LemonInTheDark replied:
did you fix this? the issue is cables not ever knowing who's actually connected to them, and just scanning constantly

Mothblocks commented on code/__HELPERS/lighting.dm:
```diff
@@ -1,9 +1,29 @@
 	appearance.color = GLOB.emissive_color
 	return appearance
 
+/// This is a semi hot proc, so we micro it. saves maybe 150ms
```
this isnt really a doc comment just an implementation detail

Mothblocks commented:
conflix

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into there-was-no-worth-in-the-fashion

LemonInTheDark committed:
renames some things, cleans some comments

LemonInTheDark committed:
Adds a setter for panel_open, so I can update_appearance on it
